maptime/main.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"log"
+	"path"
+	"strings"
+
+	"golang.org/x/tools/go/loader"
+
+	"github.com/kisielk/gotool"
+)
+
+type callbackFunc func(loc token.Position, keyStr, valStr string)
+
+func main() {
+	flag.Parse()
+	importPaths := gotool.ImportPaths(flag.Args())
+	if len(importPaths) == 0 {
+		flag.Usage()
+		return
+	}
+
+	handleImportPaths(importPaths, func(position token.Position, keyStr, valStr string) {
+		fmt.Printf("%s: Reconsider use of map[%s]%s . Storing an instance of time.Time as part of a map key is not recommended.\n", position.String(), keyStr, valStr)
+	})
+}
+
+func handleImportPaths(importPaths []string, callback callbackFunc) {
+	fs := token.NewFileSet()
+	var conf loader.Config
+	conf.Fset = fs
+
+	for _, importPath := range importPaths {
+		conf.Import(importPath)
+	}
+
+	prog, err := conf.Load()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, pkg := range prog.InitialPackages() {
+		for _, file := range pkg.Files {
+			ast.Walk(newASTVisitor(fs, pkg.Types, callback), file)
+		}
+	}
+}
+
+func newASTVisitor(fs *token.FileSet, types map[ast.Expr]types.TypeAndValue, callback callbackFunc) astVisitor {
+	return astVisitor{
+		fs:    fs,
+		types: types,
+		callback: func(position token.Position, keyStr, valStr string) {
+			callback(position, path.Base(keyStr), path.Base(valStr))
+		},
+	}
+}
+
+type astVisitor struct {
+	fs       *token.FileSet
+	types    map[ast.Expr]types.TypeAndValue
+	callback callbackFunc
+}
+
+func (v astVisitor) Visit(node ast.Node) ast.Visitor {
+	mapNode, ok := node.(*ast.MapType)
+	if !ok {
+		return v
+	}
+
+	mapType := v.types[mapNode].Type.(*types.Map)
+	position := v.fs.Position(mapNode.Map)
+	keyStr := mapType.Key().String()
+	valStr := mapType.Elem().String()
+
+	// Detects map[time.Time]<T>
+	if strings.Contains(keyStr, "time.Time") {
+		v.callback(position, keyStr, valStr)
+		return nil
+	}
+
+	// Detects map[timeAlias]<T>
+	structType, ok := mapType.Key().Underlying().(*types.Struct)
+	if ok {
+		// VERSION <= go 1.8.X
+		if structType.Field(0).Name() == "sec" &&
```

Comment:
Should probably check struct field length before indexing into the fields and expecting them to be valid?

Reply:
you're right, this is broken. I guess it just never panic'd even when I ran it on the whole m3db codebase is because we don't have any map[struct{}] which would be weird anyways. I'll fix it and add a test

Reply:
Actually it should have panic'd if the struct had less than three fields...I guess its just a very specific case. Anyways I'll fix it